src/qmozgrabresult.cpp

```diff
@@ -0,0 +1,210 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "qopenglwebpage.h"
+#include "qmozgrabresult.h"
+
+#include <QCoreApplication>
+#include <QDebug>
+#include <QOpenGLFunctions_ES2>
+#include <QPointer>
+#include <QSharedPointer>
+#include <QWindow>
+
+// Handle web page grab readiness through event loop so that connection type to the ready signal doesn't matter.
+const QEvent::Type Event_WebPageGrab_Completed = static_cast<QEvent::Type>(QEvent::registerEventType());
+
+QImage gl_read_framebuffer(const QRect &rect) {
+    QSize size = rect.size();
+    int x = rect.x();
+    int y = rect.y();
+
+    while (glGetError());
+
+    QImage img(size, QImage::Format_RGB32);
+    GLint fmt = GL_BGRA_EXT;
+    glReadPixels(x, y, size.width(), size.height(), fmt, GL_UNSIGNED_BYTE, img.bits());
+    if (!glGetError())
+        return img.mirrored();
+
+    QImage rgbaImage(size, QImage::Format_RGBX8888);
+    glReadPixels(x, y, size.width(), size.height(), GL_RGBA, GL_UNSIGNED_BYTE, rgbaImage.bits());
+    return rgbaImage.mirrored();
+}
+
+class QMozGrabResultPrivate
+{
+public:
+    QMozGrabResultPrivate(QMozGrabResult *q)
+        : q_ptr(q)
+        , ready(false)
+    {
+    }
+
+    static QMozGrabResult *create(QOpenGLWebPage *webPage, const QSize &targetSize);
+
+    QMozGrabResult *q_ptr;
+    QPointer<QOpenGLWebPage> webPage;
+    QSize textureSize;
+    QImage image;
+    Qt::ScreenOrientation orientation;
+    bool ready;
+};
+
+QMozGrabResult::~QMozGrabResult()
+{
+    if (d_ptr) {
+        delete d_ptr;
+        d_ptr = 0;
+    }
+}
+
+/*!
+ * \qmlproperty variant QMozGrabResult::image
+ *
+ * This property holds the pixel results from a grab in the
+ * form of a QImage.
+ */
+QImage QMozGrabResult::image() const
+{
+    Q_D(const QMozGrabResult);
+    return d->image;
+}
+
+bool QMozGrabResult::isReady() const
+{
+    Q_D(const QMozGrabResult);
+    return d->ready;
+}
+
+bool QMozGrabResult::saveToFile(const QString &fileName)
+{
+    Q_D(QMozGrabResult);
+    return d->image.save(fileName);
+}
+
+bool QMozGrabResult::event(QEvent *e)
+{
+    Q_D(QMozGrabResult);
+    if (e->type() == Event_WebPageGrab_Completed) {
+        d->ready = true;
+        Q_EMIT ready();
+        return true;
+    }
+    return QObject::event(e);
+}
+
+void QMozGrabResult::render(const QRect &rect)
```

Comment:
Maybe this should be called something else than render as it's more like read pixels.